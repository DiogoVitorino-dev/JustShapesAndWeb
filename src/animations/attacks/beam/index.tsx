import React, { useEffect } from "react";
import { useWindowDimensions } from "react-native";
import Animated, {
  Easing,
  SharedValue,
  cancelAnimation,
  runOnJS,
  useAnimatedStyle,
  useDerivedValue,
  useSharedValue,
  withDelay,
  withTiming,
} from "react-native-reanimated";

import { RunnableAnimation } from "@/animations/animations.type";
import Colors from "@/constants/Colors";
import { Angle, Position, Size } from "@/constants/commonTypes";
import Rectangle, { RectangleData } from "@/models/geometric/rectangle";

type BeamDirection = "vertical" | "horizontal";

export interface BeamConfig extends Partial<Position> {
  /**
   * @DocMissing
   */
  length?: number;

  /**
   * @DocMissing
   */
  size?: number;

  /**
   * @DocMissing
   */
  direction?: BeamDirection;

  /**
   * @DocMissing
   */
  attackDuration?: number;

  /**
   * @DocMissing
   */
  attackSpeed?: number;

  /**
   * @DocMissing
   */
  angle?: Angle;

  /**
   * @DocMissing
   */
  prepareDuration?: number;
  /**
   * @DocMissing
   */
  delay?: number;
}

export interface BeamProps extends BeamConfig, RunnableAnimation {}

type BeamAttack = (props: BeamProps) => React.JSX.Element;

export const Beam: BeamAttack = ({
  length,
  size = 100,
  direction = "horizontal",
  attackSpeed = 100,
  prepareDuration = 2000,
  attackDuration = 1500,
  angle = 0,
  delay = 0,
  x = 0,
  y = 0,
  start,
  onFinish,
}) => {
  const window = useWindowDimensions();

  const indicatorDimension: Size = {
    width: direction === "horizontal" ? length || window.width : size,
    height: direction === "vertical" ? length || window.height : size,
  };

  const holding = useSharedValue(0);
  const beamOpacity = useSharedValue(0);
  const indicatorOpacity = useSharedValue(0);
  const width = useSharedValue<number>(
    direction === "horizontal" ? length || window.width : 0,
  );
  const height = useSharedValue<number>(
    direction === "vertical" ? length || window.height : 0,
  );
  const collidable = useSharedValue(false);

  const beam = useDerivedValue<RectangleData>(() => ({
    width: width.value,
    height: height.value,
    angle,
    x,
    y,
    collidable: collidable.value,
  }));

  const endAnimation = (dimension: SharedValue<number>) => {
    "worklet";
    collidable.value = false;
    holding.value = 0;
    indicatorOpacity.value = 0;
    beamOpacity.value = withTiming(0, { duration: prepareDuration / 2 });

    dimension.value = withTiming(0, { duration: prepareDuration / 2 }, (fin) =>
      fin && onFinish ? runOnJS(onFinish)() : undefined,
    );
  };

  const holdAnimation = (dimension: SharedValue<number>) => {
    "worklet";
    holding.value = withTiming(1, { duration: attackDuration }, (fin) =>
      fin ? endAnimation(dimension) : undefined,
    );
  };

  const attack = (dimension: SharedValue<number>) => {
    "worklet";

    collidable.value = true;
    beamOpacity.value = 1;

    dimension.value = withTiming(
      size,
      { duration: attackSpeed, easing: Easing.out(Easing.ease) },
      (fin) => (fin ? holdAnimation(dimension) : undefined),
    );
  };

  const startAnimation = (dimension: SharedValue<number>) => {
    "worklet";

    // Prepare
    if (delay) {
      beamOpacity.value = withDelay(delay, withTiming(0.7, { duration: 0 }));
      indicatorOpacity.value = withDelay(
        delay,
        withTiming(0.2, { duration: 0 }),
      );

      dimension.value = withDelay(
        delay,
        withTiming(size / 2, { duration: prepareDuration }, (prepared) =>
          prepared ? attack(dimension) : undefined,
        ),
      );
    } else {
      beamOpacity.value = withTiming(0.7, { duration: 0 });
      indicatorOpacity.value = withTiming(0.2, { duration: 0 });

      dimension.value = withTiming(
        size / 2,
        { duration: prepareDuration },
        (prepared) => (prepared ? attack(dimension) : undefined),
      );
    }
  };

  const cancel = (dimension: SharedValue<number>) => {
    cancelAnimation(holding);
    cancelAnimation(collidable);
    cancelAnimation(beamOpacity);
    cancelAnimation(indicatorOpacity);
    endAnimation(dimension);
  };

  const startByDirection = () => {
    "worklet";
    switch (direction) {
      case "vertical":
        startAnimation(width);
        break;

      default:
        startAnimation(height);
        break;
    }
  };

  const cancelByDirection = () => {
    "worklet";
    switch (direction) {
      case "vertical":
        cancel(width);
        break;

      default:
        cancel(height);
        break;
    }
  };

  useEffect(() => {
    if (start) startByDirection();
    return () => {
      if (start) {
        cancelByDirection();
      }
    };
  }, [start]);

  const beamStyle = useAnimatedStyle(() => ({
    opacity: beamOpacity.value,
    zIndex: 10,
  }));

  const indicatorStyle = useAnimatedStyle(() => ({
    width: indicatorDimension.width,
    height: indicatorDimension.height,
    top: y,
    left: x,
    position: "absolute",
    transform: [{ rotate: angle + "deg" }],
    opacity: indicatorOpacity.value,
    backgroundColor: Colors.entity.enemy,
  }));

  return (
    <>
      <Animated.View testID="beamIndicator" style={indicatorStyle} />
      <Rectangle data={beam} style={beamStyle} />
    </>
  );
};
