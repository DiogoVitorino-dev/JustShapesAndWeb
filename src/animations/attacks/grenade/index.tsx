import React from "react";

import GrenadeFragment, { GrenadeFragmentProps } from "./grenadeFragment";

import { View } from "@/components/shared";
import { Position } from "@/constants/commonTypes";
import { AnglesUtils } from "@/utils/angleUtils";

export type GrenadeFragmentOptions = Omit<
  GrenadeFragmentProps,
  "angleDirection"
>;

export interface GrenadeProps
  extends Partial<Position>,
    GrenadeFragmentOptions {
  /**
   * @DocMissing
   */
  fragments?: number;

  /**
   * @DocMissing
   */
  rotate?: number;
}

export function Grenade({
  fragments = 5,
  rotate = 0,
  onFinish,
  duration,
  ...fragmentProps
}: GrenadeProps) {
  const createFragments = () => {
    const fragmentsView: React.JSX.Element[] = [];
    let index = fragments;
    const partialAngle = 360 / index;

    const { shiftAngle } = AnglesUtils;

    while (index > 0) {
      fragmentsView.push(
        <GrenadeFragment
          {...fragmentProps}
          onFinish={index === 1 ? onFinish : undefined}
          duration={duration}
          angleDirection={shiftAngle(partialAngle * index, rotate)}
          key={index}
        />,
      );
      index--;
    }

    return fragmentsView;
  };

  return <View>{createFragments()}</View>;
}
