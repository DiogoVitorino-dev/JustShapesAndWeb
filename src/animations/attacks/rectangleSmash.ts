import { useWindowDimensions } from "react-native";
import {
  Easing,
  WithTimingConfig,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from "react-native-reanimated";

import { RunnableAnimation, StylizedAnimation } from "../animations.type";

import { RectangleSize } from "@/models/geometric/rectangle";

export type SmashDirection = "horizontal" | "vertical";

export interface RectangleSmashConfig {
  smashTo?: SmashDirection;
  prepareDuration?: number;
  prepareAmount?: number;
}
// values in ms
enum InitialValues {
  prepareDuration = 2000,
  prepareAmount = 70,
  attackDuration = 150,
}

export interface RectangleSmashAnimation
  extends RunnableAnimation,
    StylizedAnimation {}

export function useRectangleSmashAnimation(
  size: RectangleSize,
  smashConfig?: RectangleSmashConfig,
): RectangleSmashAnimation {
  const { width, height } = useWindowDimensions();
  const prepareValue =
    smashConfig?.prepareAmount || InitialValues.prepareAmount;

  const prepareTimingConfig: WithTimingConfig = {
    duration: smashConfig?.prepareDuration || InitialValues.prepareDuration,
    easing: Easing.out(Easing.ease),
  };

  const attackTimingConfig: WithTimingConfig = {
    duration: InitialValues.attackDuration,
    easing: Easing.exp,
  };

  const opacity = useSharedValue(0.5);

  const animatedStyle = useAnimatedStyle(() => ({
    width: size.value.width,
    height: size.value.height,
    opacity: opacity.value,
  }));

  const verticalAttack = () => {
    size.value = withTiming(
      {
        width: size.value.width,
        height: size.value.height + prepareValue,
      },
      prepareTimingConfig,
      (finished) => attack(finished),
    );

    function attack(ready?: boolean) {
      if (ready) {
        size.value = withTiming(
          { width: size.value.width, height },
          attackTimingConfig,
        );
      }
    }
  };

  const horizontalAttack = () => {
    opacity.value = 1;
    size.value = withTiming(
      {
        width: size.value.width + prepareValue,
        height: size.value.height,
      },
      prepareTimingConfig,
      (finished) => attack(finished),
    );

    function attack(ready?: boolean) {
      if (ready) {
        size.value = withTiming(
          { width, height: size.value.height },
          attackTimingConfig,
        );
      }
    }
  };

  const run = () => {
    switch (smashConfig?.smashTo) {
      case "vertical":
        verticalAttack();
        break;

      default:
        horizontalAttack();
        break;
    }

    return { animatedStyle };
  };

  return { animatedStyle, run };
}
