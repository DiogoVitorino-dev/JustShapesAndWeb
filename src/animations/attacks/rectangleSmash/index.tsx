import { useEffect } from "react";
import { useWindowDimensions } from "react-native";
import Animated, {
  Easing,
  SharedValue,
  cancelAnimation,
  runOnJS,
  useAnimatedStyle,
  useDerivedValue,
  useSharedValue,
  withDelay,
  withTiming,
} from "react-native-reanimated";

import { RunnableAnimation } from "@/animations/animations.type";
import Colors from "@/constants/Colors";
import { AnimatedProps, Position } from "@/constants/commonTypes";
import Rectangle, { RectangleData } from "@/models/geometric/rectangle";
import { Collidable } from "@/scripts/collision/collisionDetector";

export type SmashDirection = "vertical" | "horizontal";

export interface RectangleSmashConfig extends Partial<Position> {
  /**
   * @DocMissing
   */
  initialLength?: number;

  /**
   * @DocMissing
   */
  size?: number;

  /**
   * @DocMissing
   */
  smashTo?: SmashDirection;

  /**
   * @DocMissing
   */
  prepareDuration?: number;

  /**
   * @DocMissing
   */
  prepareAmount?: number;

  /**
   * @DocMissing
   */
  attackSpeed?: number;

  /**
   * @DocMissing
   */
  attackDuration?: number;

  /**
   * @DocMissing
   */
  hideDuration?: number;

  /**
   * @DocMissing
   */
  delay?: number;
}

export interface RectangleSmashProps
  extends RectangleSmashConfig,
    RunnableAnimation {
  style?: AnimatedProps<"View">["style"];
}

export function RectangleSmash({
  size = 80,
  initialLength = 0,
  prepareAmount = 70,
  prepareDuration = 2000,
  attackDuration = 500,
  attackSpeed = 150,
  hideDuration = 200,
  smashTo = "horizontal",
  x = 0,
  y = 0,
  delay = 0,
  onFinish,
  start,
  style,
}: RectangleSmashProps) {
  const window = useWindowDimensions();

  const width = useSharedValue(smashTo === "horizontal" ? initialLength : size);
  const height = useSharedValue(smashTo === "vertical" ? initialLength : size);

  const opacity = useSharedValue(0);
  const opacityIndicator = useSharedValue(0);
  const collidable = useSharedValue<Collidable["collidable"]>({
    enabled: true,
    ignore: true,
  });

  const rect = useDerivedValue<RectangleData>(() => ({
    width: width.value,
    height: height.value,
    x,
    y,
    collidable: { ...collidable.value },
  }));

  const holdAnimation = (dimension: SharedValue<number>) => {
    "worklet";
    setTimeout(() => {
      endAnimation(dimension);
    }, attackDuration);
  };

  const SMASH = (dimension: SharedValue<number>) => {
    "worklet";
    const dimensionLimit =
      smashTo === "vertical" ? window.height : window.width;

    dimension.value = withTiming(
      dimensionLimit,
      {
        duration: attackSpeed,
        easing: Easing.exp,
      },
      (fin) => (fin ? holdAnimation(dimension) : undefined),
    );
  };

  const prepare = (dimension: SharedValue<number>) => {
    "worklet";
    collidable.value = { ...collidable.value, ignore: false };
    if (delay) {
      opacity.value = withDelay(delay, withTiming(1, { duration: 100 }));
      opacityIndicator.value = withDelay(
        delay,
        withTiming(0.2, { duration: 100 }),
      );

      dimension.value = withDelay(
        delay,
        withTiming(
          initialLength + prepareAmount,
          { duration: prepareDuration, easing: Easing.out(Easing.ease) },
          (finished) => (finished ? SMASH(dimension) : undefined),
        ),
      );
    } else {
      opacity.value = withTiming(1, { duration: 100 });
      opacityIndicator.value = withTiming(0.2, { duration: 100 });

      dimension.value = withTiming(
        initialLength + prepareAmount,
        { duration: prepareDuration, easing: Easing.out(Easing.ease) },
        (finished) => (finished ? SMASH(dimension) : undefined),
      );
    }
  };

  const endAnimation = (dimension: SharedValue<number>) => {
    "worklet";
    collidable.value = { ...collidable.value, ignore: true };
    opacityIndicator.value = 0;

    dimension.value = withTiming(0, {
      duration: hideDuration,
      easing: Easing.out(Easing.ease),
    });

    opacity.value = withTiming(0, { duration: hideDuration }, (fin) =>
      fin && onFinish ? runOnJS(onFinish)() : undefined,
    );
  };

  const cancel = (dimension: SharedValue<number>) => {
    "worklet";
    cancelAnimation(dimension);
    cancelAnimation(opacity);
    cancelAnimation(opacityIndicator);
    endAnimation(dimension);
  };

  useEffect(() => {
    if (start) {
      if (smashTo === "horizontal") prepare(width);
      else prepare(height);
    }
    return () => {
      if (start) {
        if (smashTo === "horizontal") {
          cancel(width);
        } else {
          cancel(height);
        }
      }
    };
  }, [start]);

  const indicatorStyle = useAnimatedStyle(() => ({
    width: smashTo === "horizontal" ? window.width : size,
    height: smashTo === "vertical" ? window.height : size,
    opacity: opacityIndicator.value,
    backgroundColor: Colors.entity.enemy,
    position: "absolute",
    top: y,
    left: x,
  }));

  return (
    <>
      <Animated.View style={indicatorStyle} />
      <Rectangle data={rect} style={[{ opacity }, style]} />;
    </>
  );
}
