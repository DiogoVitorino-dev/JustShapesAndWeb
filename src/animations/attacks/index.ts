import { Beam } from "./beam";
import { Grenade } from "./grenade";
import { RectangleSmash } from "./rectangleSmash";

export const AnimatedAttacks = {
  Grenade,
  RectangleSmash,
  /**
   * @DocMissing
   */
  Beam,
};
