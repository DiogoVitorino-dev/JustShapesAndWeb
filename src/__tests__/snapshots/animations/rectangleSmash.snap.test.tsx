import { act, render, waitFor } from "@testing-library/react-native";

import { RectangleSmash } from "@/animations/attacks/rectangleSmash";

describe("Grenade Fragment - Snapshot animation test", () => {
  beforeEach(() => jest.useFakeTimers());
  afterEach(() => jest.useRealTimers());

  it("should render correctly", async () => {
    const tree = render(<RectangleSmash start={false} />).toJSON();
    await waitFor(() =>
      act(() => {
        jest.advanceTimersByTime(5000);
      }),
    );
    expect(tree).toMatchSnapshot();
  });
});
