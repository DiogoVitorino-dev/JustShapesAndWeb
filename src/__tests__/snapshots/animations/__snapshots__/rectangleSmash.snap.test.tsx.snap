// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grenade Fragment - Snapshot animation test should render correctly 1`] = `
[
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "backgroundColor": "rgb(255, 64, 20)",
          "height": 80,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "width": 750,
        },
      }
    }
    style={
      {
        "backgroundColor": "rgb(255, 64, 20)",
        "height": 80,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 750,
      }
    }
  />,
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "height": 80,
          "left": 0,
          "top": 0,
          "transform": [
            {
              "rotate": "0deg",
            },
          ],
          "width": 0,
        },
      }
    }
    style={
      {
        "backgroundColor": "rgb(255, 64, 20)",
        "height": 80,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "transform": [
          {
            "rotate": "0deg",
          },
        ],
        "width": 0,
      }
    }
    testID="rectangleModel"
  />,
  ";",
]
`;
